libdnf5/logger/rotating_file_logger.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf5/logger/rotating_file_logger.hpp"
+
+#include "libdnf5/utils/bgettext/bgettext-mark-domain.h"
+
+#include <fcntl.h>
+#include <fmt/format.h>
+#include <string.h>
+#include <sys/file.h>
+#include <unistd.h>
+
+#include <mutex>
+
+namespace libdnf5 {
+
+class RotatingFileLogger::Impl {
+public:
+    explicit Impl(const std::filesystem::path & base_file_path, std::size_t max_bytes, std::size_t backup_count);
+
+    void write(const char * line) noexcept;
+
+private:
+    bool should_rotate(int fd, std::size_t msg_len) const noexcept;
+
+    const std::filesystem::path base_file_path;
+    const std::size_t max_bytes;
+    const std::size_t backup_count;
+
+    std::mutex stream_mutex;
+};
+
+
+RotatingFileLogger::Impl::Impl(
+    const std::filesystem::path & base_file_path, std::size_t max_bytes, std::size_t backup_count)
+    : base_file_path{base_file_path},
+      max_bytes{max_bytes},
+      backup_count{backup_count} {
+    if (::open(
+            base_file_path.c_str(), O_WRONLY | O_APPEND | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH) ==
+        -1) {
+        throw FileSystemError(errno, base_file_path, M_("Cannot open log file"));
+    }
+}
+
+
+void RotatingFileLogger::Impl::write(const char * line) noexcept {
+    try {
+        // required for thread safety
+        std::lock_guard<std::mutex> guard(stream_mutex);
+
+        auto line_len = strlen(line);
+        while (true) {
+            // open (create) the base log file and lock it
+            auto fd = ::open(
+                base_file_path.c_str(),
+                O_WRONLY | O_APPEND | O_CREAT | O_CLOEXEC,
+                S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+            if (fd == -1) {
+                return;
+            }
+            ::flock(fd, LOCK_EX);
```

Comment:
What about to use/extend the existing `libdnf5::utils::Locker`?

Reply:
I thought about it. But our `libdnf5::utils::Locker` works differently. If the file cannot be locked, an exception is thrown. We use that elsewhere. But for logs, I want the logger to wait a moment and then write the message instead of generating an exception. So either I would call `libdnf5::utils::Lock::write_lock` in a cycle (polling) or extend our Locker. In addition, it opens a file with different attributes than I would like in the case of a logger.
And now I look in the source codes and there is a bug in the implementation of `libdnf5::utils::Locker`.

- So `libdnf5::utils::Locker` would be best fixed
- and then maybe expand it so that it can be better used in more places, for example even for a logger.

Reply:
Got it. Please create an issue for the bug you've found. We can do the potential refactorings later.